.github/workflows/main.yml

```diff
@@ -0,0 +1,559 @@
+        /usr/local/osquery-toolchain/usr/bin/clang --version
+
+    - name: Select the build job count
+      shell: bash
+      id: build_job_count
+      run: |
+        echo ::set-output name=VALUE::$(($(nproc) + 1))
+
+    - name: Select the build options for the tests
+      shell: bash
+      id: tests_build_settings
+      run: |
+        if [[ "${{ matrix.build_type }}" == "RelWithDebInfo" ]] ; then
+          echo ::set-output name=VALUE::OFF
+        else
+          echo ::set-output name=VALUE::ON
+        fi
+
+    - name: Select the debug symbols options
+      shell: bash
+      id: debug_symbols_settings
+      run: |
+        if [[ "${{ matrix.build_type }}" == "Debug" ]] ; then
+          echo ::set-output name=VALUE::OFF
+        else
+          echo ::set-output name=VALUE::ON
+        fi
+
+    # When we spawn in the container, we are root; create an unprivileged
+    # user now so that we can later use it to launch the normal user tests
+    - name: Create a non-root user
+      if: matrix.build_type != 'RelWithDebInfo'
+      id: unprivileged_user
+      run: |
+        useradd -m -s /bin/bash unprivileged_user
+        echo ::set-output name=NAME::unprivileged_user
+
+    # Due to how the RPM packaging tools work, we have to adhere to some
+    # character count requirements in the build path vs source path.
+    #
+    # Failing to do so, will break the debuginfo RPM package.
+    #
+    # Symbolic links were being dereferenced, so the `mount -o bind` approach
+    # was chosen instead. The `path` parameter introduced in the checkout@v2
+    # action could easily solve this problem, but we lose access to the `.git`
+    # folder so it's not an option.
+    - name: Setup the build paths
+      id: build_paths
+      run: |
+        mkdir -p workspace/usr/src/debug/osquery/build \
+                 workspace/ccache \
+                 workspace/padding-required-by-rpm-packages/src
+
+        chown -R ${{ steps.unprivileged_user.outputs.NAME }}:${{ steps.unprivileged_user.outputs.NAME }} .
+
+        mount -o bind "$(pwd)" workspace/padding-required-by-rpm-packages/src
+
+        echo ::set-output name=SOURCE::$(realpath workspace/padding-required-by-rpm-packages/src)
+        echo ::set-output name=BINARY::$(realpath workspace/usr/src/debug/osquery/build)
+        echo ::set-output name=CCACHE::$(realpath workspace/ccache)
+
+    # One of the tests in the test suit will spawn a Docker container
+    # using this socket. Allow the unprivileged user we created
+    # to access it.
+    - name: Update the Docker socket permissions
+      if: matrix.build_type != 'RelWithDebInfo'
+      run: |
+        chmod 666 /var/run/docker.sock
+
+    - name: Update the cache (ccache)
+      uses: actions/cache@v2
+      with:
+        path: ${{ steps.build_paths.outputs.CCACHE }}
+
+        key: |
+          ccache_Linux_${{ matrix.build_type }}_${{ github.sha }}
+
+        restore-keys: |
+          ccache_Linux_${{ matrix.build_type }}
```

Comment:
This seems weird. Don't we want to list multiple restore keys? And do we want to cache with the github tag, and not just the sha?

Reply:
The way it works is that it first tries a perfect match with the key in the `key` property, then it tries a fuzzy match with the fixed prefix in the `restore-keys` (so the variable part is the sha), by preferring the newest cache first.
Using the sha the idea is that instead of overwriting the same cache entry, each new build with different commits creates a new cache entry that can be used by the worker.
That should already match the cache that contains the cached object that are closest to the build happening.
Although the key doesn't contain any separator for the branch the PR is in, and although caches have multiple scopes, when the retrieval happen the PR branch scope is preferred first.
Finally the fixed part of the cache restore key will always be present in the caches.
In fact reading here: https://docs.github.com/en/free-pro-team@latest/actions/guides/caching-dependencies-to-speed-up-workflows#using-the-cache-action (which is the same as for Azure Pipeliens in the end), I'm not sure why they make an example with multiple keys when all of them support partial matches **and** there's no other key which will ever have a shorter prefix.

Reply:
Copying from the docs:
```
The cache action first searches for cache hits for `key` and `restore-keys` in the branch containing the workflow run. If there are no hits in the current branch, the cache action searches for `key` and `restore-keys` in the parent branch and upstream branches.

You can provide a list of restore keys to use when there is a cache miss on `key`. You can create multiple restore keys ordered from the most specific to least specific. The cache action searches for `restore-keys` in sequential order. When a key doesn't match directly, the action searches for keys prefixed with the restore key. If there are multiple partial matches for a restore key, the action returns the most recently created cache.
```

So this is probably fine. We could get fancy with restore-keys, or wait till we get it working